### PR TITLE
property: create property names more eagerly.

### LINKED ISSUE #15218: User property names do not work as expected
We have some built-in property names like "fips" and "provider". It is supposed to be possible for third party provider authors to specify their own property names. Such user property names are only valid if they have a "." in the name, e.g. "foo.bar".

However attempting to use these user properties does not work as expected. If you specify them before the first fetch in which the core sees them then they fail to match, e.g. if you set the default properties to "foo.bar=yes", that will only work if you've already fetched an algorithm that had the "foo.bar" property defined for it.

Full reproducer code below:

```` C
#include <openssl/core.h>
#include <openssl/core_dispatch.h>
#include <openssl/core_names.h>
#include <openssl/provider.h>
#include <openssl/crypto.h>
#include <openssl/evp.h>

static OSSL_FUNC_provider_query_operation_fn testprov_query;
static OSSL_FUNC_digest_get_params_fn tmpmd_get_params;
static OSSL_FUNC_digest_digest_fn tmpmd_digest;

#define MYPROPERTIES "foo.bar=yes"

int tmpmd_get_params(OSSL_PARAM params[])
{
    OSSL_PARAM *p = NULL;

    p = OSSL_PARAM_locate(params, OSSL_DIGEST_PARAM_BLOCK_SIZE);
    if (p != NULL && !OSSL_PARAM_set_size_t(p, 1))
        return 0;

    p = OSSL_PARAM_locate(params, OSSL_DIGEST_PARAM_SIZE);
    if (p != NULL && !OSSL_PARAM_set_size_t(p, 1))
        return 0;

    return 1;
}

int tmpmd_digest(void *provctx, const unsigned char *in, size_t inl,
                 unsigned char *out, size_t *outl, size_t outsz)
{
    return 0;
}

const OSSL_DISPATCH testprovmd_functions[] = {
    { OSSL_FUNC_DIGEST_GET_PARAMS, (void (*)(void))tmpmd_get_params },
    { OSSL_FUNC_DIGEST_DIGEST, (void (*)(void))tmpmd_digest },
};

static const OSSL_ALGORITHM testprov_digests[] = {
    { "testprovmd", MYPROPERTIES, testprovmd_functions },
    { NULL, NULL, NULL }
};

static const OSSL_ALGORITHM *testprov_query(void *provctx,
                                          int operation_id,
                                          int *no_cache)
{
    *no_cache = 0;
    switch (operation_id) {
    case OSSL_OP_DIGEST:
        return testprov_digests;
    }
    return NULL;
}

static const OSSL_DISPATCH testprov_dispatch_table[] = {
    { OSSL_FUNC_PROVIDER_QUERY_OPERATION, (void (*)(void))testprov_query },
    { 0, NULL }
};

int testprov_provider_init(const OSSL_CORE_HANDLE *handle,
                           const OSSL_DISPATCH *in,
                           const OSSL_DISPATCH **out,
                           void **provctx)
{
    *provctx = (void *)handle;
    *out = testprov_dispatch_table;
    return 1;

}

#define DEFAULT_PROPS_FIRST       0
#define DEFAULT_PROPS_AFTER_LOAD  1
#define DEFAULT_PROPS_AFTER_FETCH 2

int props_and_providers(int propsorder)
{
    OSSL_LIB_CTX *libctx = OSSL_LIB_CTX_new();
    OSSL_PROVIDER *testprov;
    EVP_MD *testprovmd;

    OSSL_PROVIDER_add_builtin(libctx, "testprov", testprov_provider_init);

    if (propsorder == DEFAULT_PROPS_FIRST)
        EVP_set_default_properties(libctx, MYPROPERTIES);

    testprov = OSSL_PROVIDER_load(libctx, "testprov");

    if (propsorder == DEFAULT_PROPS_AFTER_LOAD)
        EVP_set_default_properties(libctx, MYPROPERTIES);

    testprovmd = EVP_MD_fetch(libctx, "testprovmd", NULL);

    if (propsorder == DEFAULT_PROPS_AFTER_FETCH) {
        EVP_set_default_properties(libctx, MYPROPERTIES);
        if (testprovmd == NULL)
            testprovmd = EVP_MD_fetch(libctx, "testprovmd", NULL);
    }

    if (testprovmd != NULL)
        printf("Success\n");
    else
        printf("Failure\n");

    EVP_MD_free(testprovmd);
    OSSL_PROVIDER_unload(testprov);
    OSSL_LIB_CTX_free(libctx);
}

int main(void)
{
    printf("Set properties first:\n");
    props_and_providers(DEFAULT_PROPS_FIRST);
    printf("Set properties after provider load:\n");
    props_and_providers(DEFAULT_PROPS_AFTER_LOAD);
    printf("Set properties after first fetch:\n");
    props_and_providers(DEFAULT_PROPS_AFTER_FETCH);

    return 0;
}
````

The above reproducer outputs the following:

````
Set properties first:
Failure
Set properties after provider load:
Failure
Set properties after first fetch:
Success
````

### PR DESCRIPTION
**Pull request by paulidale**
User defined property names were not created before the first fetch.  The rationale for this was to only maintain the user names defined by providers.  This was intended to prevent malicious memory use attacks.

Not being able to specify a default query before the first fetch is wrong.  This changes the behaviour of the property query parsing to always create property names.

Fixes #15218

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
So we basically give up on the idea of not adding property names from queries? There is one more case where parse_name is called with 0, should that be changed as well to allow to create?

**Review by t-j-h — APPROVED** [review]

File: test/user_property_test.c [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            && !TEST_true(EVP_set_default_properties(libctx, MYPROPERTIES)))
+        goto err;
+
+    if (!TEST_ptr(testprovmd = EVP_MD_fetch(libctx, "testprovmd", NULL)))
+        goto err;
+
+    if (propsorder == DEFAULT_PROPS_AFTER_FETCH) {
+        if (!TEST_true(EVP_set_default_properties(libctx, MYPROPERTIES)))
+            goto err;
+        if (testprovmd == NULL
```
Comment (mattcaswell) [inline test/user_property_test.c:115]: This is actually dead code now since you added the "goto err" in the earlier fetch.

Comment (paulidale) [reply test/user_property_test.c]: Fixed.

File: test/user_property_test.c [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    if (propsorder == DEFAULT_PROPS_AFTER_LOAD
+            && !TEST_true(EVP_set_default_properties(libctx, MYPROPERTIES)))
+        goto err;
+
+    if (!TEST_ptr(testprovmd = EVP_MD_fetch(libctx, "testprovmd", NULL)))
+        goto err;
+
+    if (propsorder == DEFAULT_PROPS_AFTER_FETCH)
+        if (!TEST_true(EVP_set_default_properties(libctx, MYPROPERTIES))
+                || !TEST_ptr(testprovmd = EVP_MD_fetch(libctx, "testprovmd",
```
[the remark this answers is not in the document]
Comment (paulidale) [reply test/user_property_test.c]: Fixed this and a second leak.

File: test/user_property_test.c [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+static int testprov_provider_init(const OSSL_CORE_HANDLE *handle,
+                                  const OSSL_DISPATCH *in,
+                                  const OSSL_DISPATCH **out,
+                                  void **provctx)
+{
+    *provctx = (void *)handle;
+    *out = testprov_dispatch_table;
+    return 1;
+
```
Comment (mattcaswell) [inline test/user_property_test.c:76]: Spurious blank line

**Review by mattcaswell — APPROVED** [review]
Approved assuming nit is fixed.

**Comment by paulidale** [comment]
See if I've got it now.

File: test/user_property_test.c [hunk 4]
```diff
@@ -45,6 +45,7 @@ static int tmpmd_digest(void *provctx, const unsigned char *in, size_t inl,
 static const OSSL_DISPATCH testprovmd_functions[] = {
     { OSSL_FUNC_DIGEST_GET_PARAMS, (void (*)(void))tmpmd_get_params },
     { OSSL_FUNC_DIGEST_DIGEST, (void (*)(void))tmpmd_digest },
+    { 0, NULL }
```
Comment (mattcaswell) [inline test/user_property_test.c:48]: Oops. Mea culpa.

**Review by mattcaswell — APPROVED** [review]
Approved assuming the CIs are ok with it this time.

**Comment by paulidale** [comment]
Merged to master.  Thanks for the feedback and reviews.

**Pull request closed without merging**